src/java.base/share/classes/javax/crypto/KEMSpi.java

```diff
@@ -0,0 +1,245 @@
+ * decapsulator. The result of either {@link #engineNewDecapsulator} or
+ * {@link #engineNewDecapsulator} must be of a single configuration,
+ * where its {@link EncapsulatorSpi#engineSecretSize()} and
+ * {@link EncapsulatorSpi#engineEncapsulationSize()} are constants.
+ * <p>
+ * A {@code KEMSpi} implementation should be immutable. It should be safe to
+ * call multiple {@code engineNewEncapsulator} and {@code engineNewDecapsulator}
+ * methods at the same time.
+ * <p>
+ * {@code EncapsulatorSpi} and {@code DecapsulatorSpi} implementations should
+ * be immutable. It should be safe to invoke multiple {@code encapsulate} and
+ * {@code decapsulate} methods at the same time. Each invocation of
+ * {@code encapsulate} should generate a new shared secret and key
+ * encapsulation message.
+ * <p>
+ * For example,
+ * {@snippet lang = java:
+ * public static class MyKEMImpl implements KEMSpi {
+ *
+ *     @Override
+ *     public KEMSpi.EncapsulatorSpi engineNewEncapsulator(PublicKey pk,
+ *             AlgorithmParameterSpec spec, SecureRandom secureRandom)
+ *             throws InvalidAlgorithmParameterException, InvalidKeyException {
+ *         if (!checkPublicKey(pk)) {
+ *             throw new InvalidKeyException("unsupported key");
+ *         }
+ *         if (!checkParameters(spec)) {
+ *             throw new InvalidAlgorithmParameterException("unsupported params");
+ *         }
+ *         return new MyEncapsulator(pk, spec, secureRandom);
+ *     }
+ *
+ *     class MyEncapsulator implements KEMSpi.EncapsulatorSpi {
+ *         MyEncapsulator(PublicKey pk, AlgorithmParameterSpec spec, SecureRandom secureRandom){
+ *             this.spec = spec != null ? spec : getDefaultParameters();
+ *             this.secureRandom = secureRandom != null ? secureRandom : getDefaultSecureRandom();
+ *             this.pk = pk;
+ *         }
+ *
+ *         @Override
+ *         public KEM.Encapsulated encapsulate(int from, int to, String algorithm) {
+ *             // ...
+ *         }
+ *
+ *         // ...
+ *     }
+ *
+ *     // ...
+ * }
+ * }
+ *
+ * @see KEM
+ * @since 21
+ */
+public interface KEMSpi {
+
+    /**
+     * The KEM encapsulator implementation, generated by
+     * {@link #engineNewEncapsulator}.
+     *
+     * @see KEM.Encapsulator
+     */
+    interface EncapsulatorSpi {
+        /**
+         * The key encapsulation function.
+         * <p>
+         * An implementation must support the case where {@code from} is 0,
+         * {@code from} is the same as the output of {@code secretSize()},
+         * and {@code algorithm} is "Generic".
+         * <p>
+         * The caller of this method has already validated the parameters to
+         * ensure that {@code algorithm} is not {@code null}, and the values
+         * of {@code from} and {@code to} are within the correct range.
+         * Therefore an implementation of this method does not need to
+         * validate them.
```

Comment:
The KEM caller does validate the parameters, but the caller may be more widely other than the KEM.   Then, the statement here could be wrong at that time.

Reply:
You can make it a required part of the specification.  But it is a error-prone design.